src/sentry/interfaces/stacktrace.py

```diff
@@ -750,9 +750,20 @@ def get_hash(self, platform=None, system_frames=True):
             if len(frames) / float(total_frames) < 0.10:
                 return []
 
+        if not frames:
+            return []
+
         output = []
-        for frame in frames:
-            output.extend(frame.get_hash(platform))
+
+        # stacktraces that only differ by the number of recursive calls should
+        # hash the same, so we squash recursive calls by comparing each frame
+        # to the previous frame
+        output.extend(frames[0].get_hash(platform))
+        prev_frame = frames[0]
+        for frame in frames[1:]:
+            if not frame == prev_frame:
```

Comment:
`!=` Brett.

Reply:
Can we reliably compare two `Frame` objects? I'm not sure if we have anything to make sure this is actually correct with the right `__eq__` stuff.

Otherwise, can we just compare against the hashes instead of the `Frames` themselves?

Reply:
Good call, but hash won't be enough. As seen in the test we apparently emit `'io.sentry.example.Application', 'recurFunc'` for both calls in `recurFunc` even though they come from different line numbers.

Python is just comparing all the attrs, right? It felt safe to me in that worst case some weird attr means we do the old thing, but if a frame is exactly the same (as in my test example which comes from a real SDK) it is skipped. Do you think I should override `__eq__` manually?

Reply:
I'd prefer we supply an explicit `__eq__` in that case, but I'm not entirely sure what the default behavior is without it. It might compare the `__dict__`, but I'd want to be sure.

Reply:
IMO, equality for the sake of `get_hash` and "are two `Frame` objects equal" are two different things and overriding `__eq__` for this single case seems like the wrong move.

> in that worst case some weird extra field on one object means we do the old thing

One of those extra fields is `vars` which are probably going to be different in recursive calls.

Reply:
Yeah, I don't know enough here. But worst case, is things that are equal compare as not equal and we get false negatives here, which doesn't make the situation worse.

Reply:
To be clear, I wasn't suggesting we shouldn't improve the equality check here (we should.)

Looking at this again, I think what would make sense here is to compare the result of `frame.get_hash` for the current and previous (instead of the frame objects themselves, which have a lot of attributes like `vars` that aren't going to be utilized in the hash construction anyway.)

Reply:
Forgot about `vars`, you're right that `__eq__` is wrong here.

> Looking at this again, I think what would make sense here is to compare the result of `frame.get_hash` for the current and previous

Matt said that too, but the issue with this is seen in the test: it would omit all but the first call to `recurFunc` even though the last frame of `recurFunc` is actually on a different line number.

So do you think I should just make a `check_recursive` function that explicitly tests equality of module, file, function, line number or whatever?

Reply:
> So do you think I should just make a check_recursive function that explicitly tests equality of module, file, function, line number or whatever? Or should we just go with the hash?

I think that makes most sense personally.

Reply:
Oops, I should have read the history more carefully — sorry about that.